### PR TITLE
#4630 Upgrade Scala 2.12.11 to 2.13.2

### PR DESCRIPTION
**Pull request by chia7712**
### Checklist
- [x] [Review Contribution Guidelines](https://ohara.readthedocs.io/en/latest/contributing.html)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#ohara-team))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #4630)
- [ ] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -292,8 +292,14 @@ subprojects {
       // see https://github.com/scala/scala/blob/2.12.x/src/compiler/scala/tools/nsc/settings/Warnings.scala#L60
       "-Xlint",
       "-Ywarn-dead-code",
-      "-Ywarn-unused-import",
-      "-Xfatal-warnings"
+      "-Ywarn-unused:imports",
```
Comment (chia7712) [inline build.gradle:295]: add more checks to unused code :)

File: docker/broker.dockerfile [hunk 2]
```diff
@@ -29,12 +29,11 @@ RUN yum install -y \
 # WARN: Please don't change the value of KAFKA_DIR
 ARG KAFKA_DIR=/opt/kafka
 ARG KAFKA_VERSION=2.5.0
-ARG SCALA_VERSION=2.12
```
Comment (chia7712) [inline docker/broker.dockerfile:32]: we don't support scala 2.12 anymore so I remove this argument

File: docker/configurator.dockerfile [hunk 3]
```diff
@@ -25,7 +25,6 @@ ARG COMMIT=$BRANCH
 ARG REPO="https://github.com/oharastream/ohara.git"
 ARG BEFORE_BUILD=""
 ARG KAFKA_VERSION=2.5.0
-ARG SCALA_VERSION=2.12
```
Comment (chia7712) [inline docker/configurator.dockerfile:28]: we don't support scala 2.12 anymore so I remove this argument

File: docker/worker.dockerfile [hunk 4]
```diff
@@ -24,12 +24,11 @@ LABEL stage=$STAGE
 # WARN: Please don't change the value of KAFKA_DIR
 ARG KAFKA_DIR=/opt/kafka
 ARG KAFKA_VERSION=2.5.0
-ARG SCALA_VERSION=2.12
```
Comment (chia7712) [inline docker/worker.dockerfile:27]: we don't support scala 2.12 anymore so I remove this argument

File: ohara-client/src/main/scala/oharastream/ohara/client/configurator/v0/JsonRefiner.scala [hunk 5]
```diff
@@ -28,20 +28,20 @@ import spray.json.{JsValue, RootJsonFormat}
   * at both url and payload, and both of them must go through the same name string check.
   * @tparam T object
   */
-trait JsonFormat[T] extends RootJsonFormat[T] {
+trait JsonRefiner[T] extends RootJsonFormat[T] {
```
Comment (chia7712) [inline ohara-client/src/main/scala/oharastream/ohara/client/configurator/v0/JsonRefiner.scala:31]: akka has a class called ```JsonFormat``` also and scala 2.13 complains the duplicate naming ...

**Comment by chia7712** [comment]
retry build

**Comment by chia7712** [comment]
retry build

File: build.gradle (hunk 1 above)
Comment (vitojeng) [reply build.gradle:295]: why remove `-Xfatal-warnings` ?

File: docker/broker.dockerfile (hunk 2 above)
Comment (vitojeng) [reply docker/broker.dockerfile:32]: Why remove `SCALA_VERSION` directly instead set `SCALA_VERSION=2.13` ?

Comment (jackyoh) [reply docker/broker.dockerfile:32]: 這邊我猜是因為怕 build 的人傳錯參數，才用 hard code 的方式

File: ohara-agent/src/main/scala/oharastream/ohara/agent/Collie.scala [hunk 6]
```diff
@@ -36,7 +36,7 @@ import oharastream.ohara.metrics.BeanChannel
 import oharastream.ohara.metrics.basic.CounterMBean
 import oharastream.ohara.metrics.kafka.TopicMeter
 
-import scala.collection.JavaConverters._
```
Comment (jackyoh) [inline ohara-agent/src/main/scala/oharastream/ohara/agent/Collie.scala:39]: 這個如果有改，應該有很多測試都會改到

File: docker/broker.dockerfile (hunk 2 above)
Comment (chia7712) [reply docker/broker.dockerfile:32]: > Why remove SCALA_VERSION directly instead set SCALA_VERSION=2.13 ?

we support only 2.13 now so the argument is not used I guess :)

File: ohara-agent/src/main/scala/oharastream/ohara/agent/Collie.scala (hunk 6 above)
Comment (chia7712) [reply ohara-agent/src/main/scala/oharastream/ohara/agent/Collie.scala:39]: the import is deprecated so ...

Comment (jackyoh) [reply ohara-agent/src/main/scala/oharastream/ohara/agent/Collie.scala:39]: ok, I got it

File: docker/broker.dockerfile (hunk 2 above)
Comment (vitojeng) [reply docker/broker.dockerfile:32]: Hmmm...This argument might be have some value, IMO.
Maybe someday in the future we need testing between different scala minor version? 
e.g 2.13.2 vs 2.13.5 ?

Comment (chia7712) [reply docker/broker.dockerfile:32]: > Maybe someday in the future we need testing between different scala minor version?
e.g 2.13.2 vs 2.13.5 ?

make sense. Let me revert it :)

File: docker/shabondi.dockerfile [hunk 7]
```diff
@@ -22,13 +22,14 @@ LABEL stage=$STAGE
...
 ARG COMMIT=$BRANCH
+ARG SCALA_VERSION=2.13
 ARG REPO="https://github.com/oharastream/ohara.git"
 ARG BEFORE_BUILD=""
 WORKDIR /testpatch/ohara
 RUN git clone $REPO /testpatch/ohara
 RUN git checkout $COMMIT
 RUN if [[ "$BEFORE_BUILD" != "" ]]; then /bin/bash -c "$BEFORE_BUILD" ; fi
-RUN ./gradlew clean build -x test -PskipManager
+RUN ./gradlew clean build -x test -PskipManager  -Pscala.version=$SCALA_VERSION
```
Comment (chia7712) [inline docker/shabondi.dockerfile:32]: @vitojeng I noticed that shabondi dockerfile does not support SCALA_VERSION so I fix it as well :)

File: docker/broker.dockerfile (hunk 2 above)
Comment (chia7712) [reply docker/broker.dockerfile:32]: done

File: ohara-agent/src/main/scala/oharastream/ohara/agent/Collie.scala (hunk 6 above)
Comment (vitojeng) [reply ohara-agent/src/main/scala/oharastream/ohara/agent/Collie.scala:39]: Java converters have been changed many times in Scala libraries, so sad.

Comment (chia7712) [reply ohara-agent/src/main/scala/oharastream/ohara/agent/Collie.scala:39]: > Java converters have been changed many times in Scala libraries, so sad.

we all hate/love compatibility :)

File: docker/shabondi.dockerfile (hunk 7 above)
Comment (vitojeng) [reply docker/shabondi.dockerfile:32]: Thanks @chia7712 , it's awesome!

File: docker/configurator.dockerfile [hunk 8]
```diff
@@ -25,7 +25,7 @@ ARG COMMIT=$BRANCH
...
+ARG SCALA_VERSION=2.13
 # this argument is used to change default version of services
 ARG OHARA_VERSION=""
 WORKDIR /testpatch/ohara
```
Comment (jackyoh) [inline docker/configurator.dockerfile:31]: configurator.dockerfile 的 38 行 -Pscala.version=$SCALA_VERSION 很像也要刪除掉

Comment (chia7712) [reply docker/configurator.dockerfile:31]: 現在不刪掉了，因為要保留調整scala version的機會

Comment (jackyoh) [reply docker/configurator.dockerfile:31]: OK, 了解。

File: gradle/dependencies.gradle [hunk 9]
```diff
@@ -51,28 +51,28 @@ versions["jackson"] = project.hasProperty('jackson.version') ? project.propertie
...
   kafkaConnectRuntime: "org.apache.kafka:connect-runtime:$versions.kafka",
   kafkaConnectJson   : "org.apache.kafka:connect-json:$versions.kafka",
   kafkaStreams       : "org.apache.kafka:kafka-streams:$versions.kafka",
-  akkaTestKit        : "com.typesafe.akka:akka-testkit_${versions["scalaMinor"]}:$versions.akka",
-  akkaStream         : "com.typesafe.akka:akka-stream_${versions["scalaMinor"]}:$versions.akka",
-  akkaHttp           : "com.typesafe.akka:akka-http_${versions["scalaMinor"]}:$versions.akkaHttp",
-  akkaHttpTestKit    : "com.typesafe.akka:akka-http-testkit_${versions["scalaMinor"]}:$versions.akkaHttp",
-  akkaHttpSprayJson  : "com.typesafe.akka:akka-http-spray-json_${versions["scalaMinor"]}:$versions.akkaHttp",
-  akkaSlf4j          : "com.typesafe.akka:akka-slf4j_${versions["scalaMinor"]}:$versions.akka",
+  akkaTestKit        : "com.typesafe.akka:akka-testkit_$defaultScalaMinor:$versions.akka",
```
Comment (vitojeng) [inline gradle/dependencies.gradle:61]: Why use `defaultScalaMinor` rather than `versions["scalaMinor"]` ?

Comment (chia7712) [reply gradle/dependencies.gradle:61]: Only scala 2.13 is supported so the minor version is not changeable

**Comment by chia7712** [comment]
retry build

**Comment by chia7712** [comment]
retry doc

**Pull request merged**